File: app/src/main/java/com/test/ashish/githubissues/IssuesRepository.java
package com.test.ashish.githubissues;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.os.AsyncTask;
import android.util.Log;

import com.test.ashish.githubissues.Pojo.GithubIssuesResponse;
import com.test.ashish.githubissues.Pojo.Issues;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class IssuesRepository {

    private IssuesDao issuesDao;

    public IssuesRepository(Application application){
        IssuesDatabase  noteDatabase = IssuesDatabase.getInstance(application);
        issuesDao = noteDatabase.issuesDao();
    }

    public void insert(List<Issues> issues){
        new InsertAsyncTask(issuesDao).execute(issues);
    }
//
    public void deleteAllIssues(long time){
        new DeleteAllAsyncTask(issuesDao).execute(time);
    }

    public LiveData<List<Issues>> getAllIssues(String orgName, String repoName) {
        LiveData<List<Issues>> issues = issuesDao.getAllIssue(orgName, repoName);
        return issues;
    }

    public void fetchApi(String orgName, String repoName){

            fetchData(orgName, repoName);
    }

    public void fetchData(String orgName, String repoName){

        retrofit2.Call<List<GithubIssuesResponse>> call1 = RetrofitClient.getInstance().getApi().getAllResponse(orgName,repoName,"open");

        call1.enqueue(new Callback<List<GithubIssuesResponse>>() {

            @Override
            public void onResponse(Call<List<GithubIssuesResponse>> call, Response<List<GithubIssuesResponse>> response) {

                if (response.code() == 200) {
                    List<GithubIssuesResponse> reponse = response.body();
                    insert(convertToIssues(reponse));
                }
            }

            @Override
            public void onFailure(Call<List<GithubIssuesResponse>> call, Throwable t) {
            }

        });


        retrofit2.Call<List<GithubIssuesResponse>> call2 = RetrofitClient.getInstance().getApi().getAllResponse(orgName,repoName,"closed");


        call2.enqueue(new Callback<List<GithubIssuesResponse>>() {

            @Override
            public void onResponse(Call<List<GithubIssuesResponse>> call, Response<List<GithubIssuesResponse>> response) {
                if (response.code() == 200) {
                    List<GithubIssuesResponse> reponse = response.body();
                    insert(convertToIssues(reponse));
                }
            }

            @Override
            public void onFailure(Call<List<GithubIssuesResponse>> call, Throwable t) {

            }

        });


    }

    public List<Issues> convertToIssues(List<GithubIssuesResponse> response){
        ArrayList<Issues> issues = new ArrayList<>();
        for(int i=0;i<response.size();i++){

            GithubIssuesResponse item = response.get(i);

             Log.d("repoUrl",item.getRepositoryUrl());
             String[] repoUrlArray = item.getRepositoryUrl().replace("https://api.github.com/repos","").split("/");

            String pathcUrl = "";
            if(item.getPullRequest() != null){
                 pathcUrl = String.valueOf(item.getPullRequest().getPatchUrl());
            }

            issues.add(new Issues(item.getNumber(), item.getTitle(), item.getUser().getLogin(),
                    pathcUrl,repoUrlArray[1], repoUrlArray[2], item.getState()));
        }
        return issues;
    }

    public static class InsertAsyncTask extends AsyncTask<List<Issues>, Void,Void> {

        private IssuesDao issuesDao;

        public InsertAsyncTask(IssuesDao issuesDao){
            this.issuesDao = issuesDao;
        }

        @Override
        protected Void doInBackground(List<Issues>... notes) {
            issuesDao.insert(notes[0]);
            return null;
        }
    }

    public static class DeleteAllAsyncTask extends AsyncTask<Long, Void,Void>{

        private IssuesDao issuesDao;

        public DeleteAllAsyncTask(IssuesDao issuesDao){
            this.issuesDao = issuesDao;

        }

        @Override
        protected Void doInBackground(Long... now) {
            issuesDao.deleteAllNotes(now[0]);
            return null;
        }
    }

}

File: app/src/main/java/com/test/ashish/githubissues/IssuesRepository.kt
package com.test.ashish.githubissues

import com.test.ashish.githubissues.IssuesDatabase.Companion.getInstance
import android.app.Application
import androidx.lifecycle.MutableLiveData
import com.test.ashish.githubissues.Pojo.Issues
import androidx.lifecycle.LiveData
import com.test.ashish.githubissues.Pojo.GithubIssuesResponse
import android.os.AsyncTask
import android.util.Log
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.ArrayList

class IssuesRepository(application: Application?) {
    private val issuesDao: IssuesDao?
    private var status: MutableLiveData<Boolean>? = null
    fun insert(issues: List<Issues?>?) {
        InsertAsyncTask(issuesDao).execute(issues)
    }

    //
    fun deleteAllIssues(time: Long) {
        DeleteAllAsyncTask(issuesDao).execute(time)
    }

    fun getAllIssues(orgName: String, repoName: String): LiveData<List<Issues?>?>? {
        val issues = issuesDao!!.getAllIssue(orgName, repoName)
        if (issues!!.value == null) {
            fetchApi(orgName, repoName)
        }
        return issues
    }

    val currentStatus: MutableLiveData<Boolean>
        get() {
            if (status == null) {
                status = MutableLiveData()
            }
            return status!!
        }

    fun fetchApi(orgName: String, repoName: String) {
        fetchData(orgName, repoName)
    }

    fun fetchData(orgName: String, repoName: String) {
        Log.d("responseBody", "$orgName : $repoName")
        val call1 = RetrofitClient.getInstance().api.getAllResponse(orgName, repoName, "all")
        call1!!.enqueue(object : Callback<List<GithubIssuesResponse?>?>{
            override fun onResponse(
                call: Call<List<GithubIssuesResponse?>?>?,
                response: Response<List<GithubIssuesResponse?>?>?
            ) {
                Log.d("responseBody", response?.body().toString())
                if (response?.code() == 200) {
                    val reponse = response.body()
                    if (reponse?.size!! > 0) {
                        insert(convertToIssues(reponse as List<GithubIssuesResponse>))
                    } else {
                        status!!.setValue(false)
                    }
                } else {
                    status!!.setValue(false)
                }
            }

            override fun onFailure(call: Call<List<GithubIssuesResponse?>?>?, t: Throwable?) {
                status!!.setValue(false)
            }

        })


        val call2 = RetrofitClient.getInstance().api.getAllResponse(orgName, repoName, "closed")
        call2!!.enqueue(object : Callback<List<GithubIssuesResponse?>?>{
            override fun onResponse(
                call: Call<List<GithubIssuesResponse?>?>?,
                response: Response<List<GithubIssuesResponse?>?>?
            ) {
                if (response!!.code() == 200) {
                    val reponse = response.body()
                    insert(convertToIssues(reponse as List<GithubIssuesResponse>))
                    if (reponse.size > 0) {
                        insert(convertToIssues(reponse))
                    } else {
                        status!!.setValue(false)
                    }
                } else {
                    status!!.setValue(false)
                }
            }

            override fun onFailure(call: Call<List<GithubIssuesResponse?>?>?, t: Throwable?) {
                status!!.setValue(false)
            }

        })
    }

    fun convertToIssues(response: List<GithubIssuesResponse>): List<Issues?> {
        val issues = ArrayList<Issues?>()
        for (i in response.indices) {
            val item = response[i]
            val repoUrlArray =
                item.repositoryUrl!!.replace("https://api.github.com/repos", "").split("/")
                    .toTypedArray()
            var pathcUrl = ""
            if (item.pullRequest != null) {
                pathcUrl = item.pullRequest!!.patchUrl.toString()
            }
            issues.add(
                Issues(
                    item.number!!, item.title!!, item.user!!.login!!,
                    pathcUrl, repoUrlArray[1], repoUrlArray[2], item.state!!
                )
            )
        }
        return issues
    }

    class InsertAsyncTask(private val issuesDao: IssuesDao?) :
        AsyncTask<List<Issues?>?, Void?, Void?>() {
        override fun doInBackground(vararg params: List<Issues?>?): Void? {
            issuesDao!!.insert(params[0])
            return null
        }
    }

    class DeleteAllAsyncTask(private val issuesDao: IssuesDao?) : AsyncTask<Long?, Void?, Void?>() {
        override fun doInBackground(vararg params: Long?): Void? {
            params[0]?.let { issuesDao!!.deleteAllNotes(it) }
            return null
        }

    }

    init {
        val noteDatabase = getInstance(application!!)
        issuesDao = noteDatabase!!.issuesDao()
    }
}